### DISCUSSION TITLE
Talking to Hardware/Devices

### DISCUSSION THREAD

**Discussion by Robatronic** [dls-controls/pythonSoftIOC#87] [Q&A] [answered]
There are existing drivers for talking to devices in EPICS. Can I use those in PythonIOC? Or is communicating with hardware (such as PLCs) something that needs to be built?

**Comment by coretl** [comment, marked as answer]
There are two answers to this:
1. If the device support is already written in Python (like https://github.com/PandABlocks/PandABlocks-client) then an IOC can be made using it. There is still a reasonable amount of work to be done (mapping each Device command to a relevant record and on_update method)
2. If there is already EPICS support, it would have to be compiled against _the exact same_ version of epics base that pythonSoftIOC uses, and then loaded. This is tricky to arrange, and is discussed in https://github.com/dls-controls/pythonSoftIOC/issues/6. At the moment, only devIocStats is compiled into pythonSoftIOC, and I would be reluctant to add anything else without a clear roadmap of where we're going with that. I would suggest it is better to manage existing EPICS support by making a traditional EPICS IOC, then adding any python logic as a separate IOC on top.
